src/app/model/Missions.test.ts

```diff
@@ -0,0 +1,155 @@
+import { Mission, MissionStatus } from "./schema";
+import missions from "./Missions";
+
+describe("Missions", () => {
+  describe("#volunteerForMission", () => {
+    const missionId = "1234";
+    let mission;
+
+    beforeEach(() => {
+      mission = new Mission();
+      mission.missionId = missionId;
+      mission.volunteerId = "";
+      mission.status = MissionStatus.unassigned;
+    });
+
+    it("assigns volunteer if missionId exists", async () => {
+      const volunteerId = "aabbbccc";
+      const mockFindById = jest.fn().mockResolvedValue(mission);
+      const mockUpdate = jest.fn().mockResolvedValue(mission);
+      const baseRepo = {
+        findById: mockFindById,
+        update: mockUpdate,
+      };
+      jest.spyOn(missions, "repo").mockReturnValue(baseRepo);
```

Comment:
Does the spy work with if you DRY-out all the setup with something similar to:
```
function mockBaseRepo({ findByIdReturn, updateReturn }) {
  const mockFindById = jest.fn().mockResolvedValue(findByIdReturn);
  const mockUpdate = jest.fn().mockResolvedValue(updateReturn);
  const baseRepo = {
    findById: mockFindById,
    update: mockUpdate,
  };
  jest.spyOn(missions, "repo").mockReturnValue(baseRepo);

  return {
    mockFindById,
    mockUpdate,
  };
}

// usage:
 
it("assigns volunteer if missionId exists", async () => {
  const volunteerId = "aabbbccc";
  const { mockFindById, mockUpdate } = mockBaseRepo({
    findByReturn: mission,
    updateReturn: mission,
  });

  await missions.volunteerForMission(missionId, volunteerId);

  expect(mockFindById).toBeCalledWith(missionId);
  expect(mockUpdate).toBeCalledWith(mission);
  expect(mission.volunteerId).toBe(volunteerId);
  expect(mission.status).toBe(MissionStatus.assigned);
});
```

Reply:
I actually prefer to scope the spies within the specific tests. The number of times I had to face problem due to timing issues with spy set up is too high! It might be tempting to DRY the code a little but I feel it makes the tests fragile.

Reply:
> 
> 
> Does the spy work with if you DRY-out all the setup with something similar to:
> 
> ```
> function mockBaseRepo({ findByIdReturn, updateReturn }) {
>   const mockFindById = jest.fn().mockResolvedValue(findByIdReturn);
>   const mockUpdate = jest.fn().mockResolvedValue(updateReturn);
>   const baseRepo = {
>     findById: mockFindById,
>     update: mockUpdate,
>   };
>   jest.spyOn(missions, "repo").mockReturnValue(baseRepo);
> 
>   return {
>     mockFindById,
>     mockUpdate,
>   };
> }
> 
> // usage:
>  
> it("assigns volunteer if missionId exists", async () => {
>   const volunteerId = "aabbbccc";
>   const { mockFindById, mockUpdate } = mockBaseRepo({
>     findByReturn: mission,
>     updateReturn: mission,
>   });
> 
>   await missions.volunteerForMission(missionId, volunteerId);
> 
>   expect(mockFindById).toBeCalledWith(missionId);
>   expect(mockUpdate).toBeCalledWith(mission);
>   expect(mission.volunteerId).toBe(volunteerId);
>   expect(mission.status).toBe(MissionStatus.assigned);
> });
> ```

Another option might be to put the mockRepo() code into beforeAll();

Reply:
Implemented!

Reply:
> > Does the spy work with if you DRY-out all the setup with something similar to:
> > ```
> > function mockBaseRepo({ findByIdReturn, updateReturn }) {
> >   const mockFindById = jest.fn().mockResolvedValue(findByIdReturn);
> >   const mockUpdate = jest.fn().mockResolvedValue(updateReturn);
> >   const baseRepo = {
> >     findById: mockFindById,
> >     update: mockUpdate,
> >   };
> >   jest.spyOn(missions, "repo").mockReturnValue(baseRepo);
> > 
> >   return {
> >     mockFindById,
> >     mockUpdate,
> >   };
> > }
> > 
> > // usage:
> >  
> > it("assigns volunteer if missionId exists", async () => {
> >   const volunteerId = "aabbbccc";
> >   const { mockFindById, mockUpdate } = mockBaseRepo({
> >     findByReturn: mission,
> >     updateReturn: mission,
> >   });
> > 
> >   await missions.volunteerForMission(missionId, volunteerId);
> > 
> >   expect(mockFindById).toBeCalledWith(missionId);
> >   expect(mockUpdate).toBeCalledWith(mission);
> >   expect(mission.volunteerId).toBe(volunteerId);
> >   expect(mission.status).toBe(MissionStatus.assigned);
> > });
> > ```
> 
> Another option might be to put the mockRepo() code into beforeAll();

Sorry, yes that wasn't very clear from me but I meant to have the mockRepo function on the global scope of the file. Seems like it's implemented that way anyway :)